### PR TITLE
Support password retrieval of ArgoCD Operator

### LINKED ISSUE #252: Support ArgoCD admin password retrieval for ArgoCD Operator
This topic arised from a discussion in [251](https://github.com/dag-andersen/argocd-diff-preview/issues/251#issuecomment-3439000415).

ArgoCD is using the secret `argocd-initial-admin-secret` with key password and that is what is read by [argocd.go](https://github.com/dag-andersen/argocd-diff-preview/blob/158e8abb66989bb00dcf778318fcfc55049af6d9/pkg/argocd/argocd.go#L358) currently.

When ArgoCD is deployed over the [ArgoCD Operator](https://argocd-operator.readthedocs.io/), which is the case for Openshift GitOps, the k8s secret is named <argocd-instance-name>-cluster with key admin.password ([reference official docs](https://argocd-operator.readthedocs.io/en/latest/usage/basics/#secrets)). As argocd-diff-preview is expecting also in other parts of the script that the instance is named `argocd` the secret name resolves to `argocd-cluster`.

To support the ArgoCD Operator password location implicitly (without additional config) the idea is to implement a fallback here in [argocd.go](https://github.com/dag-andersen/argocd-diff-preview/blob/158e8abb66989bb00dcf778318fcfc55049af6d9/pkg/argocd/argocd.go#L358) and fail after both locations are not available. This would be great as no additional secret would have to be created that holds a copy of the admin password and it would support the livecyle of rotating the admin password in the Operator without having to manually update the duplicate secret.

I will prepare a PR for that.

### PR DESCRIPTION
**Pull request by rklonner**
Lookup the k8s secret of ArgoCD Operator when default secret is not found. This should support users with an ArgoCD instance deployed with the ArgoCD Operator without further documentation.

Closes #252

### CODE REVIEW

**Comment by rklonner** [comment]
@dag-andersen : Looks easy on the first sight, fallback works in my test environment. 

However, I would like to have your opinion on the logging order... 

I would consider the ArgoCD Operator as the edge case, so it should not be confusing if you use the default setup. On the other hand, if the fallback works, it is also not nice to have the warning logged out in every script run... Therefore I choose to log out both failed secret retrievals in the second if statement. But here it is also the question what should be returned and catching the error with a second variable feels also wrong.

Would you have a suggestion for improvements?

File: pkg/argocd/argocd.go [hunk 1]
```diff
@@ -354,11 +354,16 @@ func (a *ArgoCDInstallation) login() error {
...
 	secret, err := a.K8sClient.GetSecretValue(a.Namespace, "argocd-initial-admin-secret", "password")
 	if err != nil {
-		log.Error().Msgf("❌ Failed to get secret: %s", err)
-		return "", fmt.Errorf("failed to get secret: %w", err)
+		secret, err_fallback = a.K8sClient.GetSecretValue(a.Namespace, "argocd-cluster", "admin.password")
+		if err_fallback != nil {
+			log.Error().Msgf("❌ Failed to get secret 'argocd-initial-admin-secret': %s", err)
+			log.Error().Msgf("❌ Trying ArgoCD Operator location... Failed to get secret 'argocd-cluster': %s", err_fallback)
+			return "", fmt.Errorf("failed to get secret: %w", err)
+		}
```
Comment (dag-andersen) [inline pkg/argocd/argocd.go]:
```suggestion
		log.Debug().Msgf("Failed to get password in 'argocd-initial-admin-secret'. Trying to get fallback password in 'argocd-cluster' secret.") 
		secret, err_fallback = a.K8sClient.GetSecretValue(a.Namespace, "argocd-cluster", "admin.password")
		if err_fallback != nil {
			log.Error().Err(err).Msgf("❌ Failed to get secret 'argocd-initial-admin-secret'")
			log.Error().Err(err_fallback).Msgf("❌ Failed to get fallback secret 'argocd-cluster'")
			return "", fmt.Errorf("failed to get secret: %w", err)
		}
```

**Review by dag-andersen — CHANGES_REQUESTED** [review]
Very good question! I like the implementation overall. I requested slight wording changes ⬇️  

I also added a `log.Debug().Msgf` statement because it’s extremely helpful for me when people post issues and share their debug log output 😄

File: pkg/argocd/argocd.go [hunk 2]
```diff
@@ -354,11 +354,17 @@ func (a *ArgoCDInstallation) login() error {
...
 	secret, err := a.K8sClient.GetSecretValue(a.Namespace, "argocd-initial-admin-secret", "password")
 	if err != nil {
-		log.Error().Msgf("❌ Failed to get secret: %s", err)
-		return "", fmt.Errorf("failed to get secret: %w", err)
+		log.Debug().Msgf("Failed to get password in 'argocd-initial-admin-secret'. Trying to get fallback password in 'argocd-cluster' secret.") 
+		secret, err_fallback = a.K8sClient.GetSecretValue(a.Namespace, "argocd-cluster", "admin.password")
+		if err_fallback != nil {
+			log.Error().Err(err).Msgf("❌ Failed to get secret 'argocd-initial-admin-secret'")
+			log.Error().Err(err_fallback).Msgf("❌ Failed to get fallback secret 'argocd-cluster'")
+			return "", fmt.Errorf("failed to get secret: %w", err)
```
[the remark this answers is not in the document]
Comment (dag-andersen) [reply pkg/argocd/argocd.go]: @rklonner 
I think returning `err` is perfectly fine :) They’ll see the fallback error in the log output anyway
